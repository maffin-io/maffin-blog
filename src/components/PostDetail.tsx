import React from 'react';
import Link from 'next/link';

import '@/css/prism.css';
import { Post } from '@/app/api/getPosts';
import TableOfContents from '@/components/TableOfContents';
import TagsWidget from './TagsWidget';

export type PostDetailProps = {
  post: Post;
};

export default function PostDetail(
  { post }: PostDetailProps,
): JSX.Element {
  return (
    <article>
      <div className="xl:divide-y xl:divide-gray-200 xl:dark:divide-gray-700">
        <header className="text-center pt-6 xl:pb-6">
          <h1 className="heading">
            {post.title}
          </h1>
        </header>

        <div
          className="pb-8 xl:grid xl:grid-cols-4 xl:gap-x-6"
          style={{ gridTemplateRows: 'auto 1fr' }}
        >
          <div className="sidenav">
            <div className="nav-section">
              <h2 className="text-xs uppercase tracking-wide text-gray-500 dark:text-gray-400">
                Table of contents
              </h2>
              <TableOfContents post={post} />
            </div>
            <div className="nav-section">
              <TagsWidget tags={post.tags} />
            </div>
            {
              !post.tags.includes('docs')
              && (
                <div className="nav-section pt-4 xl:pt-8">
                  <Link href="/">
                    &larr; Back to the blog
                  </Link>
                </div>
              )
            }
          </div>
          <div className="xl:col-span-3 xl:row-span-2 xl:pb-0">
            <div className="prose prose-img:mx-auto max-w-none pt-10 pb-8 dark:prose-dark">
              <div dangerouslySetInnerHTML={{ __html: post.content }} />
            </div>
          </div>
        </div>
      </div>
    </article>
  );
}
