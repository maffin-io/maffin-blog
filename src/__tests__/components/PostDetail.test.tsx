import { DateTime } from 'luxon';
import PostDetail from '@/components/PostDetail';
import { render } from '@testing-library/react';

import type { Post } from '@/app/api/getPosts';

jest.mock('@/components/TableOfContents', () => jest.fn(
  () => <div data-testid="TableOfContents" />,
));

jest.mock('@/components/TagsWidget', () => jest.fn(
  () => <div data-testid="TagsWidget" />,
));

describe('PostDetailTest', () => {
  it('renders post as expected', async () => {
    const post = {
      content: '<div>Content</div>',
      slug: 'my-blog-post',
      summary: 'Blog post summary TODO',
      tags: ['label1'],
      reading_time: '1 min',
      title: 'My Blog Post',
      date: DateTime.fromISO('2023-01-01'),
      author: {
        name: 'Name',
        avatar: 'https://avatar.image',
      },
    } as Post;
    const { container } = render(<PostDetail post={post} />);

    expect(container).toMatchSnapshot();
  });

  it('renders doc as expected', async () => {
    const post = {
      content: '<div>Content</div>',
      slug: 'my-doc',
      summary: 'Doc summary TODO',
      tags: ['docs'],
      reading_time: '1 min',
      title: 'My Doc',
      date: DateTime.fromISO('2023-01-01'),
      author: {
        name: 'Name',
        avatar: 'https://avatar.image',
      },
    } as Post;
    const { container } = render(<PostDetail post={post} />);

    expect(container).toMatchSnapshot();
  });
});
