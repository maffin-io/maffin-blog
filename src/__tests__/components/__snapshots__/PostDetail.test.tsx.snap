// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetailTest renders post as expected 1`] = `
<div>
  <article>
    <div
      class="xl:divide-y xl:divide-gray-200 xl:dark:divide-gray-700"
    >
      <header
        class="text-center pt-6 xl:pb-6"
      >
        <h1
          class="heading"
        >
          My Blog Post
        </h1>
      </header>
      <div
        class="pb-8 xl:grid xl:grid-cols-4 xl:gap-x-6"
        style="grid-template-rows: auto 1fr;"
      >
        <div
          class="sidenav"
        >
          <div
            class="nav-section"
          >
            <h2
              class="text-xs uppercase tracking-wide text-gray-500 dark:text-gray-400"
            >
              Table of contents
            </h2>
            <div
              data-testid="TableOfContents"
            />
          </div>
          <div
            class="nav-section"
          >
            <div
              data-testid="TagsWidget"
            />
          </div>
          <div
            class="nav-section pt-4 xl:pt-8"
          >
            <a
              href="/"
            >
              ‚Üê Back to the blog
            </a>
          </div>
        </div>
        <div
          class="xl:col-span-3 xl:row-span-2 xl:pb-0"
        >
          <div
            class="prose prose-img:mx-auto max-w-none pt-10 pb-8 dark:prose-dark"
          >
            <div>
              <div>
                Content
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </article>
</div>
`;
